File: UVA/AOAPC-BAC/Ch7/Example/UVA-524 Prime Ring Problem.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int n, kase = 1;

const int MAXN = 16 * 2;
vector<bool> is_prime(MAXN + 1, true);
void sieve() {
    is_prime[0] = is_prime[1] = false;
    for (int i = 2; i * i <= MAXN; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= MAXN; j += i) 
                is_prime[j] = false;
        }
    }
}

vector<vector<int>> ans;
vector<int> path;
void dfs(int s) {
    if (s == (1 << n) - 1) {
        if (is_prime[path.back() + path[0]])
            ans.push_back(path);
        return;
    }
    for (int i = 0; i < n; i++) {
        if (s & (1 << i)) continue;
        int x = i + 1;
        if (path.size() && !is_prime[path.back() + x]) continue;
        path.push_back(x);
        dfs(s | (1 << i));
        path.pop_back();
    }
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    sieve();
    while (cin >> n) {
        if (kase > 1) cout << endl;
        ans.clear();
        path.clear();
        path.push_back(1);
        dfs(1 << 0);
        cout << "Case " << kase++ << ":" << endl;
        for (auto &p : ans) {
            for (int i = 0; i < n; i++) cout << p[i] << " \n"[i == n - 1];
        }
    }
    return 0;
}

File: UVA/AOAPC-BAC/Ch7/Example/UVA-524 Prime Ring Problem.py
"""
多骰幾次，有時候會 TLE
"""

import sys
input = lambda: sys.stdin.readline().strip()
print = lambda val = "": sys.stdout.write(str(val) + "\n")

MAXN = 32
is_prime = [True] * (MAXN + 1)
is_prime[0] = is_prime[1] = False
for i in range(2, int(MAXN ** 0.5) + 1):
    if is_prime[i]:
        for j in range(i * i, MAXN + 1, i):
            is_prime[j] = False

kase = 1
while True:
    try:
        n = int(input().strip())
    except:
        break

    u = (1 << n) - 1
    ans = []
    path = [1] # 一定以 1 開頭
    def dfs(s):
        if s == u:
            if is_prime[path[-1] + path[0]]:
                ans.append(path[:])
            return
        for i in range(n):
            if s & (1 << i):
                continue
            x = i + 1
            if path and not is_prime[path[-1] + x]:
                continue
            path.append(x)
            dfs(s | (1 << i))
            path.pop()
    dfs(1 << 0)

    if kase > 1:
        print()
    print(f"Case {kase}:")
    for p in ans:
        print(" ".join(map(str, p)))
    kase += 1